man/pione-action.1
.TH pione\-action 1 "" "PIONE User Manual"
.SH NAME
.PP
pione\-action \- execute an action in literate action document.
.SH SYNOPSIS
.PP
\f[B]pione action\f[] [options] location action_name
.SH DESCRIPTION
.PP
\f[B]action\f[] command execute or show an action in literate action
document.
.SH OPTIONS
.TP
.B \-\-show
Show an action instead of execution.
.RS
.RE
.TP
.B \-d, \-\-directory PATH
Execute an action in the directory.
.RS
.RE
.SH EXAMPLES
.TP
.B pione action HelloWorld.md SayHello
Execute \f[C]SayHello\f[] action of \f[C]HelloWorld.md\f[] in current
directory.
.RS
.RE
.TP
.B pione action \-d /tmp HelloWorld.md SayHello
Execute \f[C]SayHello\f[] action of \f[C]HelloWorld.md\f[] in /tmp.
.RS
.RE
.TP
.B pione action HelloWorld.md SayHello \-\-show
Show \f[C]SayHello\f[] action of \f[C]HelloWorld.md\f[].
.RS
.RE
.SH REPORTING BUGS
.PP
Report bugs or feature requests to PIONE\[aq]s issue
tracker(https://github.com/pione/pione/issues).
.SH AUTHORS
Keita Yamaguchi.


History.txt
# History

## 0.3.2(2013/11/05)

* Added literate action document.
* Added `update-package-info` subcommand.
* Added `action` subcommand.
* Added `action:list` subcommand
* Added `--timeout SEC` option to pione-client command.
* Changed to stop processing if PIONE methods not found.
* PNML compiler can generate executable PIONE document.
* Fixed many bugs about parameters.
* Fixed many bugs about features.
* Fixed problem of global item `presence_notification_addresses`.
* Renamed and updated `SelectRuleByParam` package.
* Updated many example's packages.

## 0.3.1(2013/10/29)

* Fixed performance bug.
* Enable to generate profile reports of rule application.
* Added PNML compiler.
* Changed to stop processing when action rule happens error.
* Improved pione-clean command and the document, see `man pione-clean`.
* Added `pione` facade command.

## 0.3.0(2013/10/20)

* Improved multi-inheritance package system.
* Added package database.
* Added package cache system.
* Improved language parser and interpreter.
* Refactored many codes.
* Improved test coverage.

## 0.2.2(2013/07/02)

* Enabled to load remote package.
* Added pione-package command.
* Added PIONE archive(".ppg" file).
* Added git package.
* Fixed the problem about pione-client termination.
* Refactored location system.
* Added HTTPS location.
* Added examples.

## 0.2.1(2013/06/15)

* Added basic/advaneced parameter.
* Enabled to run rehearsal test.
* Imporved test coverage.

## 0.2.0(2013/05/31)

* Added initial package system.
* Added pione-val.
* Improved sequence.
* Improved message log framework.
* Added system log framework.
* Changed to unify process log.
* Added endurance test scripts.
* Added embedded FTP server.
* Reduced OS dependencies.
* Added examples of WorkflowPatterns.

## 0.1.4(2013/05/13)

* Sequence based expressions.
* Enable to distribute tasks based on parameters.
* Added many methods.
* Added annotations.
* Added remove and touch operation as data tuple operations.
* Modified update criteria.
* Added empty rule.
* Added new examples.

## 0.1.3(2013/04/25)

* Added new data expression literal "null" for accepting data nonexistence
* Added new data expression property keyword "neglect" and "care" for the update criterion about data mtime
* Changed log location to be in output location
* Changed log format to JSON
* Added pione-log command that generates agent activity log as XES format
* Use locations in tuples
* Enabled to test FTP location handling without setting up FTP server
* Added the feature to create domain information files(".domain_info")

## 0.1.2(2013/04/01)

* Added shorthand notation for setting parameters
* Added ticket concept for sequential processing
* Improved termination sequence
* Refactored parser and transformer

## 0.1.1(2013/03/20)

* Improved performance
* Added the concept of document parameter
* Added --list-params option
* Fixed bugs of features handling
* Improved comment parser

## 0.1.0(2013/02/19)

* First gem release


lib/pione/command/pione-action.rb
module Pione
  module Command
    # PioneAction is a command definition of "pione action" for executing
    # literate action.
    class PioneAction < BasicCommand
      #
      # basic informations
      #

      command_name "pione action"
      command_banner "execute an action in literate action document"
      PioneCommand.add_subcommand("action", self)

      #
      # options
      #

      use_option :color
      use_option :debug

      define_option(:domain) do |item|
        item.long = "--domain"
        item.desc = "use the domain information file"
        item.default = Location["./domain.dump"]
        item.value = lambda {|b| b}
      end

      define_option(:show) do |item|
        item.long = "--show"
        item.desc = "show the action without execution"
        item.value = lambda {|b| b}
      end

      define_option(:directory) do |item|
        item.short = "-d"
        item.long = "--directory PATH"
        item.desc = "execute in the PATH"
        item.value = lambda {|b| Location[b]}
      end

      #
      # command lifecycle: setup phase
      #

      setup :target
      setup :domain

      # Setup location of literate action document and action name.
      def setup_target
        abort("There are no literate action documents or packages.")  if @argv[0].nil?
        abort("Action name is needed.") if @argv[1].nil?
        @location = Location[@argv[0]]
        @name = @argv[1]
      end

      # Load a domain information file.
      def setup_domain
        if option[:domain].exist?
          @domain_info = System::DomainInfo.read(option[:domain])
        end
      end

      #
      # command lifecycle: execution phase
      #

      execute :exec

      # Update pacakge info files.
      def execute_exec
        if action = LiterateAction::Document.load(@location).find(@name)
          if option[:show]
            puts action.textize(@domain_info)
          else
            action.execute(@domain_info, option[:directory])
          end
        else
          abort(e.message)
        end
      end
    end
  end
end
